src/modules/compliance/src/lib/FileTreeWalk.h

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef COMPLIANCE_ITERATE_DIRECTORIES_H
+#define COMPLIANCE_ITERATE_DIRECTORIES_H
+
+#include <CommonUtils.h>
+#include <ContextInterface.h>
+#include <Indicators.h>
+#include <MmiResults.h>
+#include <Result.h>
+#include <cerrno>
+#include <dirent.h>
+#include <memory>
+#include <pwd.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+namespace compliance
+{
+namespace
+{
+constexpr size_t maxDepth = 32;
+template <typename Callable>
+Result<Status> FileTreeWalk(const std::string& path, Callable callable, BreakOnNonCompliant breakOnNonCompliant, ContextInterface& context, size_t depth)
+{
+    if (depth > maxDepth)
+    {
+        return Error("Maximum recursion depth reached");
+    }
+
+    auto* dir = opendir(path.c_str());
+    if (nullptr == dir)
+    {
+        int status = errno;
+        if (ENOENT == status)
+        {
+            return Status::Compliant;
+        }
+
+        OsConfigLogError(context.GetLogHandle(), "Failed to open directory '%s': %s", path.c_str(), strerror(status));
+        return Error("Failed to open directory '" + path + "': " + strerror(status), status);
+    }
+
+    Result<Status> result = Status::Compliant;
+    Result<Status> subResult = Status::Compliant;
+    struct dirent* entry = nullptr;
+    for (errno = 0, entry = readdir(dir); nullptr != entry; errno = 0, entry = readdir(dir))
+    {
+        if (entry->d_type == DT_DIR)
+        {
+            if (!strcmp(entry->d_name, ".") != 0 || !strcmp(entry->d_name, "..") != 0)
```

Comment:
>(!strcmp(entry->d_name, ".") != 0 || !strcmp(entry->d_name, "..") != 0) [](http://example.com/codeflow?start=15&length=71)

nit: add inner parenthesis for clarity and no dependence on operator precedence?